src/winbind.c

```diff
@@ -245,6 +247,15 @@ uint32_t winbind_srv_auth(char *user, char *domain,
 
     memcpy(ntlmv2_key->data, wbc_info->user_session_key, ntlmv2_key->length);
 
+    // num_sids is INOUT: IN - maximum number of sids to be returned
+    //                    OUT - actual number of sids
+    if (wbc_info->num_sids < *num_sids) {
+        *num_sids = wbc_info->num_sids;
+    }
+    for (uint32_t i = 0; i < *num_sids; i++) {
+        memcpy(&sids[i], &wbc_info->sids[i].sid, sizeof(ntlm_raw_sid));
+    }
+
```

Comment:
I am thinking what we may want to do here is to actually convert this structure in something like json (or other representation) so we can serialize it in a single string and keep all the information in our context w/o having to hold onto the original structure.
It will require some parsing back on the receiving end, but that is probably ok. It's done only once per session establishment.

Reply:
Yes, for the purpose of better compatibility we can convert each SID into a string and combine all them into some textual representation (like json). It will be more user-friendly and more transparent for clients who knows nothing about SIDs and binary format, but...
We just need to use some common ground with GSSAPI - usually gss_inquire_sec_context_by_oid uses binary data (for example EXTRACT_AUTHZ_DATA, INQ_SSPI_SESSION_KEY, GET_TKT_FLAGS, etc). This is low-level library so data can be raw binary. I just followed the convention of GSSAPI.

Reply:
What is returned is what applications request, in case of the SSPI Session Key, no other representation is possible or make sense, given it is a cryptographic key.
TKT Flags is just a bifield, and GSS_KRB5_EXTRACT_AUTHZ_DATA_FROM_SEC_CONTEXT_OID is used to return arbitrary Authorization Data.

But for SIDs we get to have options.
Strings sound more portable than a binary representation, otoh if the primary user always have to convert back strings to the same binary representation used internall by windbindd, then it looks like a waste of resources.

So I guess my question is: does your application use the binary representation, ir would it not care or even prefer to compare/use strings internally ?

Reply:
Yes, we do use binary representation internally (i.e. we have own version of class similar to wbcDomainSid but with many methods to handle it). So on reception of binary buffer we don't convert from wbcDomainSid but just copy binary blob by memcpy(). However, we don't actually care if SIDs are reported in any other format - we already have all methods to make conversion.